Assign cost center in posting journal

lomamech opened the pull request:
- Assign cost centers during Payroll Processes
- use LEFT JOIN for Cost_center_allocation_basis
- Update trial balance for write in cost_center_aggregate table

closes #5921

lomamech committed:
Fix sql on procedures voucher

lomamech committed:
Deleted file fee_center in HEAD

lomamech committed:
Update and Improve use table cost_center instead fee_center

lomamech committed:
Update trial balance for write in cost_center_aggregate table

lomamech committed:
feature(Assign Cost Center)
- Assing cost centers during Payroll Processes

closes #5921

lomamech committed:
Clean and sanitaze code

lomamech committed:
Clean and sanitaze code: hound

jmcameron commented on server/controllers/finance/cost_center.js:
```diff
@@ -276,6 +276,49 @@ function del(req, res, next) {
+  const sqlGetAccountNotTitle = `
+    SELECT a.id AS account_id, a.label, a.number
+    FROM account AS a
+    WHERE a.type_id <> 6;
```
It would be clearer if a constant for '6' was defined and used with an informative variable name.

jmcameron commented on server/controllers/finance/cost_center.js:
```diff
@@ -276,6 +276,49 @@ function del(req, res, next) {
+    db.exec(sqlGetAccountNotTitle),
+  ]);
+
+  accountsCostCenter.forEach(accountCostCenter => {
```
Could this be done by a more involved SQL query?

lomamech replied:
:wrench: I will try

jmcameron commented on server/controllers/finance/cost_center.js:
```diff
@@ -276,6 +276,49 @@ function del(req, res, next) {
 }
 
+// This function searches for account membership at a cost center,
+// and returns an object table that will allow when creating a transaction
```
I found this description confusing.   Could you re-write it and explain what you are trying to do in a little more detail?

jmcameron commented on server/controllers/payroll/multiplePayroll/commitment.js:
```diff
@@ -88,6 +92,36 @@ function commitments(employees, rubrics, rubricsConfig, account, projectId, user
     item => (item.is_employee !== 1 && item.is_discount === 1 && item.totals > 0),
   );
 
+  // Here we affect for the elements that will constitute the transaction
```
Did you mean 'associate' or 'assign' instead of 'affect'?

jmcameron reviewed (commented):
I looked over the code and made a few comments.

I do not understand the payroll side of Bhima very well, so my review is very limited.
I did not try to test it.

Question:  Do we need a way to determine whether all the payroll transactions and related postings to the journal have been updated and assigned a cost center?

jniles commented on server/models/migrations/next/migrate.sql:
```diff
@@ -142,3 +142,18 @@ INSERT IGNORE INTO `cost_center_allocation_basis` VALUES
+date: 2021-09-16
+description: Add column cost_center_id and principal_center_id in Voucher Item
+*/
+CALL add_column_if_missing('voucher_item', 'cost_center_id', 'MEDIUMINT(8) UNSIGNED NULL');
```
Nice :+1:

jniles commented on server/models/procedures/migration-process.sql:
```diff
@@ -144,3 +144,132 @@ BEGIN
+*/
+
+DROP PROCEDURE IF EXISTS PostVoucher;
+DELIMITER $$
```
This shouldn't be in `migration-process.sql`.  Just replace the procedure found in the `procedures/voucher.sql`  file.

lomamech replied:
Thank you I had hesitations

jniles reviewed (commented):
This looks pretty good!  Thanks for updating the trial balance with the cost center aggregate procedures.

lomamech committed:
Fix migration file use cost_center instead fee_center

lomamech committed:
improvement(Assign Cost Center)
- Remove not necessary procedure in migration-process
- Use a SQL query for get the cost center corresponding by account

lomamech committed:
test(Assign Cost Center)
- Implement unit test for check if the account has been associated with
  cost center
- Add new function AssignCostCenterParams

jmcameron commented on server/controllers/finance/cost_center.js:
```diff
@@ -276,6 +276,65 @@ function del(req, res, next) {
+    JOIN (
+    SELECT a.id AS account_id, a.label, a.number
+    FROM account AS a
+    WHERE a.type_id <> ?
```
I suggest you use ${accountTitle} here instead of passing it as a parameter to db.exec() since it will never change.

jmcameron commented on server/controllers/finance/cost_center.js:
```diff
@@ -276,6 +276,65 @@ function del(req, res, next) {
+ * This function examines in the object in parameter, the parameter
+ * account_id and checks if this account corresponds to which cost
+ * center and returns the same object with two new
+ * parameters cost_center_id and principal_center_id
```
Please explain argument: 'rubrics'.  What is it?  I'm not quite sure what 'rubic' means in this context.

And what is 'params'.  Not clear why it is plural since you seem to be using it as a singular item.  In the usage below, it seems to represent type of account.  So I suggest renaming it appropriately to make the code easier to understand.

jmcameron reviewed (commented):
As before, I have limited understanding of account structures, payroll, etc.  So my review is limited and I do not feel that I know enough to approve the PR. 

I did look over the code and made a few comments.

I tested it and the unit tests and integration tests ran fine.

lomamech committed:
Improve and add comment detailed

lomamech committed:
Clean and sanitaze code

jniles commented on server/controllers/finance/cost_center.js:
```diff
@@ -276,6 +276,71 @@ function del(req, res, next) {
+@param {String} params - This is the element with which the comparison
+      * will be made for the list of accounts by cost center
+ */
+function assignCostCenterParams(accountsCostCenter, rubrics, params) {
```
We generally use `params` to refer to an object.  Could we call this `key` or `field` or something like that?

jniles commented on server/models/procedures/trial_balance.sql:
```diff
@@ -315,6 +315,16 @@ BEGIN
   ON DUPLICATE KEY UPDATE credit = credit + VALUES(credit), debit = debit + VALUES(debit);
 
+  -- write into cost_center_aggregate
+  INSERT INTO cost_center_aggregate (
```
Very nice :+1:

jniles reviewed (approved):
This looks like a good first step to me.  I haven't tested it yet on a computer, but the code is clean and the aggregation functions will be useful.

lomamech committed:
Resolve and fix conflict with branch master

lomamech committed:
Improvement: using key instead params

jniles commented on server/controllers/finance/cost_center.js:
```diff
@@ -276,6 +276,71 @@ function del(req, res, next) {
+@param {String} key - This is the element with which the comparison
+      * will be made for the list of accounts by cost center
+ */
+function assignCostCenterParams(accountsCostCenter, rubrics, key) {
```
Good 👍

jniles reviewed (approved):
LGTM.

bors r+
